docs/wallets.md

```diff
@@ -35,14 +35,15 @@ The process for entering shares is:
-   * For substitution errors, the wallet may highlight the offending 4-character window or the offending character. It may also show the corrected character.
-   * If the wallet can determine insertion or deletion errors, it should highlight the offending 4-character window. When detecting insertion or deletion errors, the wallet may assume that the correct share length is a multiple of 16 bytes.
+   * If the header is invalid, the wallet SHOULD highlight this and request confirmation from the user before allowing additional data to be entered. An invalid header is one that starts with a character other than `0` or `2` through `9`, or one which starts with `0` but whose share index is not `S`. For shares after the first, a header is also invalid if its threshold and identifier do not match those of the first share.
+   * If the checksum is invalid, the wallet SHOULD use an error-correction algorithm to locate errors in the share and show these to the user. It MAY additinally determine corrected share data, but if so, the wallet MUST show these corrections to the user rather than silently applying them.
+   * To show locations of substitution errors, the wallet SHOULD highlight the offending 4-character window or the specific offending character.
+   * If the wallet can determine insertion or deletion errors, it SHOULD highlight the offending 4-character window or the specific location of the inserted or missing character. When detecting insertion or deletion errors, the wallet MAY assume that the correct share length is 16, 32 or (optionally) 64 bytes.
 1. If the share length is *not* 16, 32 or 64 bytes, but the checksum passes, the wallet should confirm that the user intends to import a non-standard share length.
 1. After the first share has been entered and accepted, the wallet now knows the seed ID and threshold value.
    * If the first share had index `S`, this was the actual seed and the import process is complete.
    * Otherwise, the first character of the share will be a numeric character between `2` and `9` inclusive. The user must enter this many shares in total.
+   * Wallets MAY encrypt and store recovery progress, to allow recovery without having all shares available at once. The details of this are currently outside of the scope of this specification.
 1. The user should then enter the remaining shares, in the same manner as the first.
    * The wallet may pre-fill the header (threshold value and seed ID).
    * If the user tries to repeat an already-entered share index, they should be prevented from entering additional data until it is corrected, with the exception that `?` may be used as a share index arbitrarily many times. The wallet may guide the user by indicating that a share index has been repeated; if the user indicates that they are not repeating the share, the share index should be replaced by `?`.
```

Comment:
erasure correction MUST not replace the ? with a previously valid share's index.

small performance improvement that helps insertion/deletion correction as well.

Reply:
This would make error correction harder to implement. I'd rather not recommend it.

Reply:
If it finds a valid correction that causes a share to be reused it should error out and make the user remove at least one of the shares from the set.

Brute force algorithms are easy to tweak to force certain conditions on the corrected string, versus standard error correction algorithms which will always find the closest match. If the closest match has an invalid share index, I'm very doubtful that the "next-closest match" which does have a valid share index will be unique, and even if it is, I don't know how to find it, and even if I could, I don't think that finding a worse match (in terms of distance) is a better idea than just rejecting the share data.

In other words, this situation can't happen except when the user has 5 or more errors in their string **and** there is a valid string with distance 4 or less, whose share index has already been used. So we definitely have two possible corrections (the distance-4 and the distance-5 one) and probably many more. I don't see any principled way to pick one so we should give up.

Reply:
The situation also can happen when correcting insertions and deletions since the checksum has no "4 or less" guarantee.

The correction algorithm has to consider every possible location as a potential edit. So here it can leave the pre-filled header alone.

Giving up is reasonable when multiple corrections are both valid shares and valid share set members for equal edit distance. I'd treat a "valid" corrected share that does not belong in the share set as invalid and try to do an additional round of insert/delete. That's only ~1,600x more candidates to check. The 20-bits of known ident should break any ties and produce just 0 or 1 valid correction.

Reply:
I'm going to leave this alone, other than to capitalize "should".

Let me spend some time implementing error correction in rust-codex32, and maybe trying to implement insertion/deletion correction, and then see if I develop an intuition about what sorts of constraints are reasonable for correction implementations.